cmd/kubeadm/app/preflight/checks.go

```diff
@@ -99,7 +99,12 @@ func (CRICheck) Name() string {
 
 // Check validates the container runtime through the CRI.
 func (criCheck CRICheck) Check() (warnings, errors []error) {
-	if err := criCheck.exec.Command("sh", "-c", fmt.Sprintf("crictl -r %s info", criCheck.socket)).Run(); err != nil {
+	crictlPath, err := criCheck.exec.LookPath("crictl")
+	if err != nil {
+		errors = append(errors, fmt.Errorf("unable to find command crictl: %s", err))
+		return warnings, errors
+	}
+	if err := criCheck.exec.Command("sh", "-c", fmt.Sprintf("%s -r %s info", crictlPath, criCheck.socket)).Run(); err != nil {
```

Comment:
Why we have here execution via "sh -c ... " ?
criCheck.socket is practically unverified user input, meaning, any shell expansion/functions will be executed silently by kubeadm.

Reply:
Well spotted! I think this is an existing bug that just carried along. Now is a good time to also review other uses of `sh -c` in preflight checks.